[Op#50737] Convert new storage page to primer design

akabiru opened the pull request:
https://community.openproject.org/wp/50737

- [x] Select Provider form
     - [x] Add [Docs Link](https://www.openproject.org/docs/system-admin-guide/integrations/nextcloud/)
     - [x] Retrieve provider form on select change
     - [x] Switch select caption by provider type
     - [x] TODO: Create Task to add One Drive Storage Documentation (https://community.openproject.org/wp/50983) (https://github.com/opf/openproject/pull/14140)
- [x] Nextcloud Form
     - [x] 🐛  fix automatically managed validation errors
     - [x] 🔧  oauth client should return turbo stream partials, not the full layout html
     - [x] Do not render application password on edit
- [x] One Drive Form
- [x] Update Edit view with OneDrive context
     - [x] Fix configuration check for one drive general info
     - [x] Fix clipboard copy action
- ~Manage page refresh in new storage follow, maintain state~ 
- [x] Setup Task for Primer Flash (https://community.openproject.org/wp/50985)
- [x] Checkup Mobile View
   - [x] 🐛 Long text overflows  (https://community.openproject.org/wp/51008)

### Demo

#### Select Provider 📹 

https://github.com/opf/openproject/assets/17295175/cc977b57-8867-4219-8c2d-c4782fd29c28


#### Create Nextcloud Storage 📹 

https://github.com/opf/openproject/assets/17295175/611dbe72-7be2-4ef3-9354-8874a71e51eb


#### One Drive Storage 📹 

https://github.com/opf/openproject/assets/17295175/a5788034-0b58-47b0-a221-8ccc0426f92a

akabiru committed:
chore[Op#50737]: Break up `StorageProviderForm` for easier composition

akabiru committed:
feat[Op#50737]: Expand SaveOrCancel form options

akabiru committed:
feat[Op#50737]: Expand GeneralInfo form options

akabiru committed:
fix[Op#50737]: update cancel button path

akabiru committed:
feat[Op#50737]: Expand FrameComponent to allow explicit id with context

akabiru committed:
feat[Op#50737]: Let SaveOrCancel component extend sys args directly

akabiru committed:
feat[Op#50737]: Let ProviderTypeSelect component extend sys args directly

akabiru committed:
chore[Op#50737]: Rename `SaveOrCancelForm` to `SubmitOrCancelForm`

akabiru committed:
chore[Op#50737]: Rename `StorageGeneralInfo` cmpt to just `GeneralInfo` for consistency

akabiru committed:
chore[Op#50737]: Extract button options to methods

akabiru committed:
feat[Op#50737]: Allow optional skip_provider_type_strategy validation for select provider form

akabiru committed:
feat[Op#50737]: Blueprint for select storage provider form

akabiru committed:
chore[Op#50737]: Use a turbo stream to reduce HTML sent back, no need to render the full layout

akabiru committed:
feat[Op#50737]: Decompose the StorageViewComponent

akabiru committed:
chore[Op#50737]: Rename OAuthCredentialsInfo component for consistency

akabiru committed:
chore[Op#50737]: Amend typo

akabiru committed:
feat[Op#50737]: Continue to decompose the StorageViewComponent

akabiru committed:
fix[Op#50737]: update automatically managed form for turbo stream

akabiru committed:
fix[Op#50737]: update general info update to turbo streams

akabiru committed:
fix[Op#50737]: amend oauth client render policy

akabiru committed:
fix[Op#50737]: redirect to list page upon automatic management completion

akabiru committed:
chore[Op#50737]: Update success message

akabiru committed:
fix[Op#50737]: mend edit vs new forms for NC automatic management settings

akabiru committed:
fix[Op#50737]: Remove name from select provider form, add prompt

akabiru committed:
feat[Op#50737]: Introduce OneDrive general info form

akabiru committed:
chore[Op#50737]: Update storage header and title to "New file storage"

akabiru committed:
Merge branch 'dev' into implementation/50737-convert-new-storage-page-to-primer-design

akabiru committed:
fix[Op#50737]: Remove unused form closing method check

akabiru committed:
chore[Op#50737]: fix controller name

akabiru committed:
chore[Op#50737]: rm unused edit template

akabiru committed:
feat[Op#50737]: Extend clipboard copy component to allow input group options

akabiru committed:
feat[Op#50737]: Extend OAuthClient components to account for OneDrive storage

akabiru committed:
fix[Op#50737]: Read cancel button path from method

akabiru committed:
fix[Op#50737]: configuration check inclusive of one drive

akabiru committed:
Revert "fix[Op#50737]: Remove unused form closing method check"

This reverts commit 57e43711472435007cf3961af49e5891417ae81b.

akabiru committed:
fix[Op#50737]: configuration check label

akabiru committed:
chore[Op#50737]: make label names consistent

akabiru committed:
fix[Op#50737]: scope storage page titles by feature flag

akabiru committed:
tests[Op#50737]: Multiple fixes to correct edit vs new storage flow

akabiru committed:
fix[Op#50737]: name label by section configuration check

akabiru committed:
fix[Op#50737]: control automatically managed section via options

akabiru committed:
fix[Op#50737]: Mend OAuthClient form for one drive, introduce feature tests

akabiru committed:
tests[Op#50737]: Add New File storage feature specs

akabiru committed:
feat[Op#50737]: Auto retrieve provider form on select change

akabiru committed:
fix[Op#50737]: Switch select caption by provider type

akabiru committed:
chore[Op#50737]: define provider docs links

akabiru committed:
Merge branch 'dev' into implementation/50737-convert-new-storage-page-to-primer-design

akabiru committed:
fix[Op#50737]: Fix clipboard copy by setting HTML value on clientID input

akabiru committed:
fix[Op#50737]: Use RSpec wait correctly, as a drop in replacement for expect

akabiru committed:
chore[Op#50737]: render oauth clients instructions by provider

akabiru reviewed (commented):
🚧 Bug logged for Mobile view overflow issues- to be resolved separately. https://community.openproject.org/wp/51008

HDinger commented on frontend/src/stimulus/controllers/dynamic/storages/oauth-client-form.controller.ts:
```diff
@@ -30,27 +30,53 @@
+    }
 
-    this.redirectUriTarget.value = this.redirectUri();
+    if (this.clientIdTarget.value === '' || this.clientSecretTarget.value === '') {
```
It is somehow inconsistent that you disable the button here but nowhere else. On the other places you let the user press the button and render the error message if there is a value missing. I think this would be fine here as well.

akabiru replied:
Makes sense, I went by the designs here- but I agree that consistency is the better option here 👍🏾

akabiru replied:
I spent (significant) time on this today and it's not a simple change as we'll have to rework the OAuth Clients service workflow to accommodate this behaviour. Hence, I broker it down in to a [CODE MAINTENANCE wp](https://community.openproject.org/wp/51033)

HDinger commented on modules/storages/app/components/storages/admin/automatically_managed_project_folders_info_component.html.erb:
```diff
@@ -0,0 +1,32 @@
+    end
+
+    if editable_storage?
+      grid.with_area(:"icon-button", tag: :div, color: :subtle) do
```
The color should be `:muted` accoriding to Figma. This is especially important for the high-contrast mode. (Same is true for the other icons on the right side)

akabiru replied:
Thanks for catching this! 👍🏾

HDinger commented on modules/storages/app/components/storages/admin/oauth_client_info_component.html.erb:
```diff
@@ -0,0 +1,36 @@
+    end
+
+    grid.with_area(:description, tag: :div, color: :subtle, test_selector: 'storage-oauth-client-id-description') do
+      render(Primer::Beta::Truncate.new(font_weight: :light)) { provider_oauth_client_description }
```
This looks a bit different than specified in figma due to the `font_weight`. I noticed that "automatically managed folder" form as well.

HDinger commented on modules/storages/app/components/storages/admin/oauth_client_info_component.html.erb:
```diff
@@ -0,0 +1,36 @@
+    end
+
+    if editable_storage?
+      grid.with_area(:"icon-button", tag: :div, color: :subtle) do
```
The button should have the color `:muted`

HDinger reviewed (commented):
Hi @akabiru 

great job, it works already quite nice! 🥳 @apfohl and I had a look together and only found some minor issues while clickling through:

- [x] From time to time the routing felt weird. I try to describe the cases where we expected something else:
  - [x] After I chose a type, and then press "Cancel", I expected to go back to the index page. Instead I was able to chose a type again. Further, I expected to be able to change the type afterwards in the select field. It felt weird, that this is not possible. From an accessibility point, it is also hard to communicate that the options of a select have changed.
  - [x] When closing any other block than the "general information", it is exactly reversed: I get redirected to the Index Page, although I would expect to stay on that page where I am. This is especially an issue when I try to go one step back and see the oauth credentials I already saved (see video below). This happens however only on the Nextcloud page. For Onedrive, it behaves like expected: closing a block, leads to the block being closed and there is no redirect.

https://github.com/opf/openproject/assets/7457313/7223b14a-85b5-4e56-ada6-dce724883588

  - [x] I sometines found it hard to get back the index page, as the breadcrumbs are missing and it is not intuitive that I have to press Save or Cancel in one of the boxes.. E.g. when I create a new OneDrive storage, enter the general information, but cancel the oauth credentials, I am on a page with no return.  This is however an issue on the edit pages as well. I think the breadcrumbs would solve that for all of them. (#13962)

<img width="300" alt="Bildschirmfoto 2023-11-15 um 09 38 36" src="https://github.com/opf/openproject/assets/7457313/a5d6d3a3-ea7b-4248-b656-d86cccf66026">

- [x] The box headers all have the wrong colour as they should be grey like on the index page
- [x] There are some places where the font-weight or the color does not match exactly the figma specification. I tried to point out those that I found, but maybe you could also keep an eye open for it.
- [x] From an accessibility point of view, I am wondering whether this "magical" appearance of new blocks after the type selection is catched by a screenreader 🤔 Do you know anything about whether the turbo responses are accessible out of the box? (https://community.openproject.org/wp/51021)


@apfohl and I also tried to have a look at the code, but due to the pure mass it was kind of hard for us to really get a good overview of it. IMaybe you could give us a short introdction and point out the most important parts?

akabiru committed:
fix[Op#50737]: Borderbox headers should have the fg-muted color

akabiru committed:
fix[Op#50737]: icon buttons should have the fg-muted color

akabiru committed:
fix[Op#50737]: Match font weights with design spec

akabiru committed:
fix[Op#50737]: Remove Cancel button from OAuthApplicationInfoCopy component

aaron-contreras commented on modules/storages/app/components/storages/admin/forms/oauth_client_form_component.html.erb:
```diff
@@ -2,23 +2,81 @@
+            Storages::Admin::OAuthClientForm.new(
+              form,
+              storage:,
+              client_id_input_options: {
+                data: {
+                  'storages--oauth-client-form-target': 'clientId',
+                  action: 'input->storages--oauth-client-form#toggleSubmitButtonDisabled '\
+                          'input->storages--oauth-client-form#setRedirectUriValue'
+                }
+              },
+              client_secret_input_options: {
+                data: {
+                  'storages--oauth-client-form-target': 'clientSecret',
+                  action: 'input->storages--oauth-client-form#toggleSubmitButtonDisabled'
+                }
+              }
+            )
```
Is there a reason you're explicitly specifying the `input` event as the listener here? Stimulus provides an "Event Shorthand" notation which lets you rely on default events for different elements (unless you explicitly want to listen on a different event (e.g. `change`):

https://stimulus.hotwired.dev/reference/actions#event-shorthand

<img width="931" alt="image" src="https://github.com/opf/openproject/assets/61627014/6f00ca00-b88c-48a6-8600-269f0bfb7ac2">

akabiru replied:
Just out of preference as I find the explicit definitions more clear and easier to follow

aaron-contreras replied:
Got it, thought it was some special case.

akabiru committed:
fix[Op#50737]: Mark provider type select as readonly after storage selection

akabiru committed:
tests[Op#50737]: wait for flash message after storage creation

akabiru committed:
fix[Op#50737]: More font fixes to match design

akabiru committed:
fix[Op#50737]: Make "Cancel" button routing more consistent

1. In "General information" section, cancel redirects back to the index
   page
2. All other "Cancel" buttons stay on the same page, but hide the forms

akabiru reviewed (commented):
Thank you for the detailed review @HDinger ! I've addressed majority of the issues and the larger topics assigned to work packages. As agreed, we'll go ahead and merge this first version. 🙇🏾

aaron-contreras commented:
Great work! @akabiru

akabiru commented:
> Great work! @akabiru

Thank you @aaron-contreras ! ❤️
